scripts/check_cpp_format.py

```diff
@@ -0,0 +1,133 @@
+#!/usr/bin/python
+
+import fnmatch
+import io
+import multiprocessing
+import os
+import queue
+import subprocess
+import sys
+import threading
+
+# Parse command-line arguments
+if len(sys.argv) == 1:
+    print("Usage: check_cpp_format.py [-fix] <paths>")
+    sys.exit()
+i_arg = 1
+fix = False
+if sys.argv[i_arg] == "-fix":
+    fix = True
+    i_arg += 1
+roots = []
+for root in sys.argv[i_arg:]:
+    roots.append(root)
+
+# Parse environment variables
+prune_paths = set()
+if "PRUNE_PATHS" in os.environ:
+    last_token = io.StringIO()
+    in_quotes = False
+    for ch in os.environ["PRUNE_PATHS"]:
+        if ch in ["'", '"']:
+            if last_token.tell() != 0:
+                prune_paths.add(last_token.getvalue())
+                last_token = io.StringIO()
+            in_quotes = not in_quotes
+        elif ch == " ":
+            if in_quotes:
+                last_token.write(ch)
+            elif last_token.tell() != 0:
+                prune_paths.add(last_token.getvalue())
+                last_token = io.StringIO()
+            # else tokens are separated by multiple spaces
+        else:
+            last_token.write(ch)
+    if last_token.tell() != 0:
+        prune_paths.add(last_token.getvalue())
+        last_token = io.StringIO()
+clang_format = os.environ.get("CLANG_FORMAT", "clang-format")
+
+# Hard-coded constants
+prune_names = {"build*", ".#*"}
+file_suffixes = {".cpp", ".h", ".cu", ".cuh"}
+
+
+# Create the task queue
+tasks = queue.Queue()
+
+# Launch consumer threads
+class Consumer:
+    def __init__(self):
+        self.thread = threading.Thread(target=self.entry)
+        self.check_failed = False
+
+    def entry(self):
+        while True:
+            item = tasks.get()
+            if item is None:
+                tasks.put(None)
+                return
+            if fix:
+                print("fixing ", item)
+                cmd = [clang_format, " -style=file -i ", item]
+            else:
+                cmd = [clang_format, "-style=file", "-output-replacements-xml", item]
+
+            completed = subprocess.run(cmd, capture_output=True, check=False)
+            if not fix and (completed.returncode != 0 or ("<replacement " in str(completed.stdout))):
+                self.check_failed = True
+                print(item, " NOT OK")
+
+
+consumers = []
+for i in range(multiprocessing.cpu_count()):
+    cur_cons = Consumer()
+    consumers.append(cur_cons)
+    cur_cons.thread.start()
```

Comment:
I think you can do this much more easily with `ThreadPool` (https://docs.python.org/3/library/multiprocessing.html#multiprocessing.pool.ThreadPool). Then you don't need to manage threads and queues yourself.

Reply:
It would definitely not be easier for me. I've looked at the documentation for `ThreadPool`, but can't quickly figure out what's going on.